crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,673 @@
+};
+
+use anyhow::{
+    anyhow,
+    Context,
+    Result,
+};
+
+/// Represents different formats of a balance
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BalanceVariant {
+    /// Default format: no symbol, no token_decimals
+    Default(Balance),
+    /// Denominated format: symbol and token_decimals are present
+    Denominated(DenominatedBalance),
+}
+
+#[derive(Debug, Clone)]
+pub struct TokenMetadata {
+    /// Number of token_decimals used for denomination
+    pub token_decimals: usize,
+    /// Token symbol
+    pub symbol: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DenominatedBalance {
+    value: Decimal,
+    unit: UnitPrefix,
+    symbol: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum UnitPrefix {
+    Giga,
+    Mega,
+    Kilo,
+    One,
+    Milli,
+    Micro,
+    Nano,
+}
+
+impl TokenMetadata {
+    /// Query [TokenMetadata] through the node's RPC
+    pub async fn query(client: &Client) -> Result<Self> {
+        let sys_props = client.rpc().system_properties().await?;
+
+        let default_decimals = json!(12);
+        let default_units = json!("UNIT");
+        let token_decimals = sys_props
+            .get("tokenDecimals")
+            .unwrap_or(&default_decimals)
+            .as_u64()
+            .context("error converting decimal to u64")?
+            as usize;
+        let symbol = sys_props
+            .get("tokenSymbol")
+            .unwrap_or(&default_units)
+            .as_str()
+            .context("error converting symbol to string")?;
+        Ok(Self {
+            token_decimals,
+            symbol: symbol.to_string(),
+        })
+    }
+}
+
+impl FromStr for BalanceVariant {
+    type Err = anyhow::Error;
+
+    /// Attempts to parse the balance either in plain or denominated formats
+    /// If the balance is provide without the token symbol,
+    /// then it is treated as raw.
+    /// Otherwise, the balance is attempted to be parsed in a denominated format
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let input = input.replace('_', "");
+        // if we cannot parse the balance in raw format
+        // it means it is in a denominated format
+        let result = match input.parse::<Balance>() {
+            Ok(balance) => BalanceVariant::Default(balance),
+            Err(_) => BalanceVariant::Denominated(DenominatedBalance::from_str(&input)?),
+        };
+        Ok(result)
+    }
+}
+
+impl FromStr for DenominatedBalance {
+    type Err = anyhow::Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let symbols = value
+            .trim_start_matches(|ch: char| ch.is_numeric() || ch == '.' || ch == ',');
+        let unit_char = symbols
+            .chars()
+            .next()
+            .context("no units or symbols present")?;
+        let unit: UnitPrefix = match unit_char {
+            'G' => UnitPrefix::Giga,
+            'M' => UnitPrefix::Mega,
+            'K' => UnitPrefix::Kilo,
```

Comment:
```suggestion
            'k' => UnitPrefix::Kilo,
```
https://en.wikipedia.org/wiki/International_System_of_Units#Prefixes